src/dfcx_scrapi/tools/search_util.py

```diff
@@ -510,3 +713,264 @@ def find_event_handlers(self):
+                "initial_prompt_fulfillment",
+            ]
+        ]
+        param_reprompt_event_handler_df = param_df[
+            [
+                "flow_name",
+                "page_name",
+                "parameter_name",
+                "reprompt_event_handlers",
+            ]
+        ]
+        event_handler_df = self.get_event_handler_df(
+            page_df, param_reprompt_event_handler_df
+        )
+
+        fulfillment_df = pd.concat(
+            [
+                page_df.drop(
+                    columns=[
+                        "parameters",
+                        "route_groups",
+                        "routes",
+                        "event_handlers",
+                    ]
+                ).rename(columns={"entry_fulfillment": "fulfillment"}),
+                event_handler_df.rename(
+                    columns={"trigger_fulfillment": "fulfillment"}
+                ),
+                route_df.rename(columns={"trigger_fulfillment": "fulfillment"}),
+                param_initial_prompt_fulfillment_df.rename(
+                    columns={"initial_prompt_fulfillment": "fulfillment"}
+                ),
+            ],
+            ignore_index=True,
+        ).dropna(subset=["fulfillment"], axis="index")
+        return fulfillment_df
+
+    def get_flow_df(self, agent_id: str):
+        """Gets a flow dataframe for an agent.
+
+        Args:
+            agent_id: ID of the Dialogflow CX agent.
+        Returns:
+            flow dataframe with columns:
+                flow_name,
+                flow_id,
+                routes,
+                event_handlers,
+                route_groups
+        """
+        flowlist = self.flows.list_flows(agent_id=agent_id)
+        flow_df = pd.DataFrame(
+            [
+                {
+                    "flow_name": flow.display_name,
+                    "flow_id": flow.name,
+                    "routes": flow.transition_routes,
+                    "event_handlers": flow.event_handlers,
+                    "route_groups": flow.transition_route_groups,
+                }
+                for flow in flowlist
+            ]
+        )
+        return flow_df
+
+    def get_page_df(self, flow_df: pd.DataFrame):
+        """Gets pages dataframe for an agent.
+        Args:
+            flow_df: flow dataframe from get_flow_df().
+        Returns:
+            page dataframe with columns:
+                flow_name,
+                page_name,
+                routes,
+                event_handlers,
+                route_groups,
+                parameters,
+                entry_fulfillment
+        """
+        page_df = (
+            flow_df[["flow_name", "flow_id"]]
+            .assign(page_obj=flow_df.flow_id.apply(self.pages.list_pages))
+            .explode("page_obj", ignore_index=True)
+            .assign(
+                page_name=lambda df: df.page_obj.apply(
+                    attrgetter("display_name")
```

Comment:
@Greenford this method is failing here.

Error message I'm getting is: 
```
AttributeError: 'float' object has no attribute 'display_name'
```

Reply:
@Greenford I identified and fixed this issue.  
It was an edge case where a Flow could exist, but no Pages had been created inside the Flow yet.

Reply:
@kmaphoenix Gotcha! Is there a specific agent you used to find that error? I might re-do the code for style consistency - the explode function will add an np.nan (the float in question) if one of the values in the exploded column is an empty list. I'm making a note to look for other such cases - explodes missing a following dropna